### PR TITLE
adapter: Log origin of requests

### PR DESCRIPTION
**Pull request by parker-timmerman**
### Motivation
* This PR adds a known-desirable feature.
    * #18094 

There have been several times that a cluster has come under heavy load, but we don't know where the requests are coming from. This PR begins logging the `application_name` session var to prometheus as a label in the `mz_adapter_commands` metric. We map the `application_name` to one of a few possible values:

* `unspecified`
* `unrecognized`
* `psql`
* `dbt`
* `web_console`

This way we can keep the cardinality of our prometheus metric under control, but get context as to where a request is coming from.

In addition to logging the `application_name`, this PR also updates our `WebSocketAuth` JSON message to optionally deserialize an `options` map, and our `api/sql` HTTP endpoint to deserialize an `options` map query param. And for every entry we then set the corresponding session var. This is very similar to what we do when [initially handling a pgwire connection](https://github.com/MaterializeInc/materialize/blob/main/src/pgwire/src/protocol.rs#L233).

### Tips for reviewers
* All of the changes to support the HTTP endpoint are in a single commit called `http support`. Those changes can be reviewed entirely on their own. (I wish GitHub supported stacked PRs 😞)

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by mjibson — COMMENTED** [review]
Can we use [`application_name`](https://www.postgresql.org/docs/15/runtime-config-logging.html#GUC-APPLICATION-NAME) instead? The http/ws endpoints might need to be taught how to send arbitrary startup options to do that. I think using this is better than inventing our own because it'll work with all existing postgres tools.

The prometheus metrics for this new field should, however, be similarly locked down as they are here. That is, we have a list of known values, and anything else gets unknown or other.

**Comment by parker-timmerman** [comment]
> Can we use [`application_name`](https://www.postgresql.org/docs/15/runtime-config-logging.html#GUC-APPLICATION-NAME) instead? The http/ws endpoints might need to be taught how to send arbitrary startup options to do that. I think using this is better than inventing our own because it'll work with all existing postgres tools.
> 
> The prometheus metrics for this new field should, however, be similarly locked down as they are here. That is, we have a list of known values, and anything else gets unknown or other.

100%, I didn't know that existed, I'll update the PR to use that PG connections! For queries sent to our HTTP and WS endpoints, do you think we should parse `application_name` from a query param? AFAIU the pgwire protocol is separate from those endpoints, right?

**Comment by mjibson** [comment]
Correct, the http and ws protocols are separate. However I think they should operate in a similar way which is to have some method of sending an arbitrary list of options. Query params certainly seems like they would work for http (although I'm a bit worried that we than can't use query params to specify other stuff; maybe a param named `options` where the actual vaule is a `key=value` string (clearly with the `=` encoded)? That's ugly too. Unsure. WS could get this added to its initial auth message easily.

**Comment by chaas** [comment]
> Can we use [application_name](https://www.postgresql.org/docs/15/runtime-config-logging.html#GUC-APPLICATION-NAME) instead?

Would this allow us to identify connections from dbt adapter as well? I just got a request for that and was curious if we could use the solution here for it.

**Comment by parker-timmerman** [comment]
How does this sound as an approach:

### Extraction
We'll read the Postgres [`application_name`](https://www.postgresql.org/docs/15/runtime-config-logging.html#GUC-APPLICATION-NAME) parameter from the following places:
1. The initial [pgwire Startup Message](https://github.com/MaterializeInc/materialize/blob/main/src/pgwire/src/server.rs#L123)
2. From a query param named `application_name` for requests to:
    a. [`api/sql`](https://github.com/MaterializeInc/materialize/blob/main/src/environmentd/src/http.rs#L521) (one-off HTTP requests)
    b. [`/api/experimental/sql`](https://github.com/MaterializeInc/materialize/blob/main/src/environmentd/src/http.rs#L136) (initializing a websocket)

> Would this allow us to identify connections from dbt adapter as well?

It looks like the dbt postgres adapter does set the `application_name`, [PR](https://github.com/dbt-labs/dbt-core/pull/3182), so [1] allows us to add support for dbt!

#### Notes
* Providing multiple unrelated query params in a URL is supported, so we shouldn't need to nest them all underneath a single one like `options`. Unless we wanted to define some pgwire-like `config` query param that we nest everything under?
* I see as part of `SqlRequest` there is an [`ExtendedRequest`](https://github.com/MaterializeInc/materialize/blob/main/src/environmentd/src/http/sql.rs#L195) type that includes a `params` field. We could also read the `application_name` from here, but what makes that tricky is that allows websocket users to provide a different `application_name` on each request, which seems odd?

**Comment by mjibson** [comment]
> Would this allow us to identify connections from dbt adapter as well? I just got a request for that and was curious if we could use the solution here for it.

Probably? Depends on if dbt sets its application name or not. I'd bet they do.

**Comment by mjibson** [comment]
pgwire already implements application_name. No need to do any work there. We already do specific things if its set ([example](https://github.com/MaterializeInc/materialize/blob/e4402a884199d4a1e6cb3cb4fe3aa80fa9f7d92c/src/pgwire/src/protocol.rs#L1329)).

My concern with hard coding `&application_name=blah` as a specifically supported option is that then we'd have to manually support all the other options too. We should do something more general to support arbitrary options. We could decide to use the direct query params, but that would then preclude using query params for anything else in the future. Hence my above proposal for `&options=key%3Dvalue&options=otherkey%3Dothervalue`, which only reserves a single word out of the query params.

Websocket always starts with an [auth](https://dev.materialize.com/api/rust-private/mz_environmentd/enum.WebSocketAuth.html) message, to which we can add maybe a `options: Vec<(String, String)>` (uhh maybe that's not valid JSON because of the tuple? maybe serde converts a JSON array into a tuple? dunno), no need for query params there I think.

**Comment by parker-timmerman** [comment]
Ahhh, I see what you mean now. I wanted to use query params because I thought it was the easiest way to get support for both the HTTP endpoint and the WS endpoint, but that's definitely kind of a can of worms.

I'll add an `options` field to the WS auth message like you're suggesting, JSON supports maps so we should be able to do `BTreeMap<String, String>` I think? Then we can [plumb those into session vars like we do for pgwire](https://github.com/MaterializeInc/materialize/blob/main/src/pgwire/src/protocol.rs#L232), how does that sound?

**Comment by mjibson** [comment]
SGTM

**Comment by benesch** [comment]
Also really like the idea of using `application_name` here instead of introducing a new thing!

**Comment by ggnall** [comment]
fyi @hlburak

File: src/adapter/src/client.rs [hunk 1]
```diff
@@ -483,6 +484,8 @@ impl SessionClient {
     {
         let session = self.session.take().expect("session invariant violated");
         let mut typ = None;
+        let application_name = session.vars().application_name();
```
Comment (mjibson) [inline src/adapter/src/client.rs:487]: This should maybe just be a property and method on the session since it gets used enough. It'd have to know when that variable changes though to update itself so...maybe it's fine as is?

File: src/environmentd/src/http/sql.rs [hunk 2]
```diff
@@ -99,6 +107,28 @@ async fn run_ws(state: &WsState, mut ws: WebSocket) {
         ))
         .await;
 
+    // Send any notices that might have been generated on startup.
+    let notices = client
```
Comment (mjibson) [inline src/environmentd/src/http/sql.rs:111]: Can this become a function since it's called multiple places and does some specific string munging?

File: src/environmentd/tests/server.rs [hunk 3]
```diff
@@ -957,3 +958,80 @@ fn test_mz_system_user_admin() {
...
+    let (mut ws, _resp) = tungstenite::connect(ws_url).unwrap();
+    let options = BTreeMap::from([(
+        "application_name".to_string(),
+        "billion_dollar_idea".to_string(),
+    )]);
+    util::auth_with_ws(&mut ws, Some(options));
+
+    // Query to make sure we get back the correct session var, which should be
+    // set from the options map we passed with the auth.
+    let json = "{\"query\":\"SHOW application_name;\"}";
```
Comment (mjibson) [inline src/environmentd/tests/server.rs:982]: This test LGTM, but at also reveals maybe a problem in our WS protocol. The pgwire protocol, on startup, responds with all of the session settings. We don't do that here, forcing a SHOW instead to display one. I'm not sure which is better (to auto show on startup or not), but it's a notable difference.

File: src/environmentd/tests/auth.rs [hunk 4]
```diff
@@ -1002,6 +1004,7 @@ fn test_auth_base() {
                 auth: &WebSocketAuth::Basic {
                     user: "bad user".to_string(),
                     password: frontegg_password.to_string(),
+                    options: Some(BTreeMap::default()),
```
Comment (mjibson) [inline src/environmentd/tests/auth.rs:1007]: Why are these None sometimes or Some but empty sometimes? Does the Auth struct even need to be an option? If it's a btreemap but not specified in the json, what does serde do?

File: src/environmentd/src/http/sql.rs [hunk 5]
```diff
@@ -830,3 +866,69 @@ fn is_txn_exit_stmt(stmt: &Statement<Raw>) -> bool {
...
     )
 }
+
+#[cfg(test)]
+mod tests {
+    use super::WebSocketAuth;
+
+    #[test]
+    fn smoke_test_websocket_auth_parse() {
+        let basic = "{ \"user\": \"mz\", \"password\": \"1234\" }";
```
Comment (mjibson) [inline src/environmentd/src/http/sql.rs:876]: Light preference to use `r#` syntax here for easier JSON reading and writing.

File: src/environmentd/src/http/sql.rs [hunk 6]
```diff
@@ -830,3 +866,69 @@ fn is_txn_exit_stmt(stmt: &Statement<Raw>) -> bool {
         Statement::Commit(_) | Statement::Rollback(_) | Statement::Prepare(_)
     )
 }
+
+#[cfg(test)]
+mod tests {
+    use super::WebSocketAuth;
+
+    #[test]
+    fn smoke_test_websocket_auth_parse() {
```
Comment (mjibson) [inline src/environmentd/src/http/sql.rs:875]: Could this be converted to a table-based test so it's easier to extend in the future?

File: src/environmentd/src/http/sql.rs [hunk 7]
```diff
@@ -68,8 +69,15 @@ pub async fn handle_sql_ws(
 #[derive(Serialize, Deserialize, Debug)]
 #[serde(untagged)]
 pub enum WebSocketAuth {
-    Basic { user: String, password: String },
-    Bearer { token: String },
+    Basic {
+        user: String,
+        password: String,
+        options: Option<BTreeMap<String, String>>,
```
Comment (mjibson) [inline src/environmentd/src/http/sql.rs:75]: (asked elsewhere): Does this need to be an Option, or is there a way for serde to have not-present mean empty? Is that a good idea?

File: src/environmentd/src/http.rs
Comment (mjibson) [inline src/environmentd/src/http.rs:1]: Does this file also need to teach the HTTP endpoint about query param options? The web console uses a bunch of not-websocket stuff (afaik). Am also fine with that being a followup pr.

File: src/sql/src/session/hint.rs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    /// Request came from `psql`.
+    Psql(Private),
+    /// Request came from `dbt`.
+    Dbt(Private),
+    /// Request came from our web console.
+    WebConsole(Private),
+}
+
+impl ApplicationNameHint {
+    pub fn from_str(s: Option<&str>) -> Self {
```
Comment (mjibson) [inline src/sql/src/session/hint.rs:47]: I don't think this can ever be none because it's must always exist in the session, but defaults to the empty string.

**Review by mjibson — COMMENTED** [review]
Looking really good!

**Comment by mjibson** [comment]
Forgot to note: the websocket-api user document also needs updating. Can happen in another PR.

File: src/sql/src/session/hint.rs (hunk 8 above)
Comment (parker-timmerman) [reply src/sql/src/session/hint.rs:47]: Yeah this is an artifact of a previous impl, updated!

File: src/environmentd/src/http.rs
Comment (parker-timmerman) [reply src/environmentd/src/http.rs:1]: Looking through the cloud repo we definitely still use the HTTP endpoint. I'll cover this in a followup PR since I think that's bound to be a bit more complex since we need to parse a dict from query params?

File: src/environmentd/src/http/sql.rs (hunk 7 above)
Comment (parker-timmerman) [reply src/environmentd/src/http/sql.rs:75]: There is! Updated such that if `options` is not present, we'll deserialize to an empty `BTreeMap`. IMO this is a good idea because when using the map we were already doing `options.unwrap_or_default()`, thus essentially treating no options and empty options as the same. In the future if we care about no options vs empty options we can change this

File: src/environmentd/src/http/sql.rs (hunk 6 above)
Comment (parker-timmerman) [reply src/environmentd/src/http/sql.rs:875]: Done!

File: src/environmentd/src/http/sql.rs (hunk 5 above)
Comment (parker-timmerman) [reply src/environmentd/src/http/sql.rs:876]: Good idea, that definitely makes it cleaner!

File: src/environmentd/tests/auth.rs (hunk 4 above)
Comment (parker-timmerman) [reply src/environmentd/tests/auth.rs:1007]: Originally I did this just to make sure we had coverage of both cases, but now that options is not an Option, I changed all of these to `BTreeMap::default`

File: src/environmentd/tests/server.rs (hunk 3 above)
Comment (parker-timmerman) [reply src/environmentd/tests/server.rs:982]: I feel like we should probably try to follow the pgwire protocol? I made a task to track this: https://github.com/MaterializeInc/materialize/issues/18503

File: src/environmentd/src/http/sql.rs (hunk 2 above)
Comment (parker-timmerman) [reply src/environmentd/src/http/sql.rs:111]: Good call! Updated this

File: src/adapter/src/client.rs (hunk 1 above)
Comment (parker-timmerman) [reply src/adapter/src/client.rs:487]: Makes sense to me, I made a new method on `Session` for `application_name` like we have for `user`

**Comment by parker-timmerman** [comment]
> Forgot to note: the websocket-api user document also needs updating. Can happen in another PR.

Since it looks like I'll need another PR to update the LTS docs, I included the change here

File: doc/user/content/integrations/websocket-api.md [hunk 9]
```diff
@@ -59,8 +59,9 @@ To authenticate using a username and password, send an initial text or binary me
 
 ```
 {
-  "user": "<Your email to access Materialize>"
-	"password": "<Your app password>",
+    "user": "<Your email to access Materialize>",
+    "password": "<Your app password>",
+    "options": { <Optional map of run-time settings> }
```
Comment (jseldess) [inline doc/user/content/integrations/websocket-api.md:64]: By run-time settings, do we mean https://materialize.com/docs/sql/show/#session-variables? If so, we say "Optional map of session variables"? I wonder, also, if there's a way for us to mention this in the intro to this section or under General Semantics.

File: doc/user/content/integrations/websocket-api.md [hunk 10]
```diff
@@ -69,6 +70,7 @@ To authenticate using a token, send an initial text or binary message containing
 ```
 {
 	"token": "<Your access token>",
+    "options": { <Optional map of run-time settings> }
```
Comment (mjibson) [inline doc/user/content/integrations/websocket-api.md:73]: Both of these new options fields also need to go into the typescript definition, along with some more descriptive prose and maybe an example showing how to set one. (Again, totally fine to do docs in another PR.)

File: src/environmentd/src/http.rs [hunk 11]
```diff
@@ -303,9 +315,41 @@ where
...
+            .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
+
+        // Apply options that were provided in query params.
+        let session = client.0.session();
+        let maybe_options = if params.options.is_empty() {
+            // It's possible 'options' simply wasn't provided, we don't want that to
+            // count as a failure to deserialize
+            Ok(BTreeMap::<String, String>::default())
+        } else {
+            serde_json::from_str(&params.options)
```
Comment (mjibson) [inline src/environmentd/src/http.rs:329]: Ah a json-encoded query param. My proposal had some jank in it that required url-encoding an equals sign (`?options=value%3Dkey`), but this seems better.

File: src/environmentd/src/http.rs [hunk 12]
```diff
@@ -303,9 +315,41 @@ where
...
+                const LOCAL: bool = false;
+                if let Err(err) = session.vars_mut().set(&key, VarInput::Flat(&val), LOCAL) {
+                    session.add_notice(AdapterNotice::BadStartupSetting {
+                        name: key.to_string(),
+                        reason: err.to_string(),
+                    })
+                }
+            }
+        } else {
+            // If we fail to deserialize options, emit a Notice.
```
Comment (mjibson) [inline src/environmentd/src/http.rs:343]: This should error and close the connection, not just send a notice.

File: src/environmentd/src/http.rs [hunk 13]
```diff
@@ -289,8 +292,17 @@ where
...
         req: &mut http::request::Parts,
-        _state: &S,
+        state: &S,
     ) -> Result<Self, Self::Rejection> {
+        #[derive(Debug, Default, Deserialize)]
+        struct Params {
+            #[serde(default)]
+            options: String,
+        }
+        let params: Query<Params> = Query::from_request_parts(req, state)
```
Comment (mjibson) [inline src/environmentd/src/http.rs:302]: This needs some docs at some point too.

File: src/environmentd/tests/server.rs [hunk 14]
```diff
@@ -1035,3 +1036,121 @@ fn test_ws_notifies_for_bad_options() {
...
+struct Notice {
+    message: String,
+    #[serde(rename = "severity")]
+    _severity: String,
+}
+
+#[test]
+#[cfg_attr(miri, ignore)] // too slow
+fn test_http_passes_options() {
+    let server = util::start_server(util::Config::default()).unwrap();
```
Comment (mjibson) [inline src/environmentd/tests/server.rs:1062]: Can these tests also be made a table test? They appear to do a lot of the same thing. That would avoid starting one server for each test, which is quite expensive.

**Review by mjibson — APPROVED** [review]
Looking pretty much ready to go! Some comments, but mostly I think this could merge and we can fix up docs or test stuff in followups if needed.

File: doc/user/content/integrations/websocket-api.md (hunk 9 above)
Comment (parker-timmerman) [reply doc/user/content/integrations/websocket-api.md]: Updated to say "Optional map of session variables". I'm going to further refine the docs for the websocket and http APIs in a followup PR, I'll take a pass at updating the intro section or General Semantics there :)

File: doc/user/content/integrations/websocket-api.md (hunk 10 above)
Comment (parker-timmerman) [reply doc/user/content/integrations/websocket-api.md]: Sounds good! I'll put these in a followup PR

File: src/environmentd/src/http.rs (hunk 11 above)
Comment (parker-timmerman) [reply src/environmentd/src/http.rs:329]: Yeah I've seen the equals sign pattern in other places, but the json-encoded query param seemed to be the easiest to build on the web side, and parse on the server side.

File: src/environmentd/src/http.rs (hunk 12 above)
Comment (parker-timmerman) [reply src/environmentd/src/http.rs]: Updated!

File: src/environmentd/src/http.rs (hunk 13 above)
Comment (parker-timmerman) [reply src/environmentd/src/http.rs:302]: Totally, I'll add these in a followup PR

File: src/environmentd/tests/server.rs (hunk 14 above)
Comment (parker-timmerman) [reply src/environmentd/tests/server.rs:1062]: De-duped the tests here, didn't exactly make it a table test though since the different scenarios seemed to be asserting different enough stuff that trying to unify it into a single function would add too much boiler plate. Happy to revist this if you'd like though :)
